src/python/grpcio/grpc/_cython/_cygrpc/aio/grpc_aio.pyx.pxi

```diff
@@ -13,16 +13,32 @@
 # limitations under the License.
 
 
-cdef bint _grpc_aio_initialized = 0
+cdef bint _grpc_aio_initialized = False
+# NOTE(lidiz) Theoretically, applications can run in multiple event loops as
+# long as they are in the same thread with same magic. However, I don't think
+# we should support this use case. So, the gRPC Python Async Stack should use
+# a single event loop picked by "init_grpc_aio".
+cdef object _grpc_aio_loop
```

Comment:
Not sure about this change, Asyncio runs one loop per thread. If the user spins up a new Thread, hence a new Asyncio loop this will lead to an invalid usage of call_soon - explicitly or implicitly because was called behind the scenes by a create_task- when we try to wake up a future that was created in the new loop.

There is an outstanding question about how the driver must work in a multithread and multievent loop environment, and TBH we do not need to give an answer to this right now since I would say that this is a not well spread and used pattern (the multithread and multiloop environments)

Maybe we should create a Github issue and add it into the Alpha, Beta or GA release. I would be tempted to have this solved by the Beta release.

Meanwhile, are we happy about having this pattern that might raise exceptions if the user starts using multiple threads and multiple loops? or should we add a guard for avoiding this pattern and simply raise a RuntimeError when it happens?

Reply:
> There is an outstanding question about how the driver must work in a multithread and multievent loop environment, and TBH we do not need to give an answer to this right now since I would say that this is a not well spread and used pattern (the multithread and multiloop environments)

`asyncio` functions are not allowed to invoke **from other threads** or **from another event loop**, not only powerful functions like `loop.create_task` and `loop.call_soon`, but also `Task.cancel`, `Task.done`, etc.. Otherwise, an exception will be raised by `asyncio` module:

```
RuntimeError: Non-thread-safe operation invoked on an event loop other than the current one
```

If we activate the AsyncIO IO manager, the IO operations will inevitably be invoked in other thread, which have given us a hard time to make sync stack and async stack compatible. To avoid the above `RuntimeError`, we will need to pick an event loop.

About multiple event loop objects in one thread, I don't think it is a valid use case... Each individual servers and channels can only live in one event loop, and it might take some effort to make them working at full-capacity under multi-loop environment.

> Meanwhile, are we happy about having this pattern that might raise exceptions if the user starts using multiple threads and multiple loops? or should we add a guard for avoiding this pattern and simply raise a RuntimeError when it happens?

`asyncio` module raises `RuntimeError` for us.

Reply:
I think that we are on the same page about not sharing RPC objects through different threads/loops if we eventually we support many threads and many loops, we should explicitly prescript to the users to not share objects across different loops and threads, otherwise `RuntimeErros` will start appearing.

Also, we are on the same page that the scenario in Asyncio when you have multiple threads is by having 1 loop per thread, other options are not even supported by Asyncio.

But at the moment that we are saving at the very beginning which Asyncio loop will be used during the whole application, we explicitly forbidding the pattern for having many threads and many loops, , where theoretically the gRPC library might operate - saying theoretically because has been never tested.

My point about raising our own function was for raising explicitly an exception that will tell the user that nowadays the `Aio` module does not work in a multi-thread environment, and neither a multi-loop since we are sticking to the same Asyncio loop here.

Also, since we were reducing the options for the user I was wondering if in the future - short, middle e or even a long term - we could start investigating how we could remove this restriction, so giving the chance of running the `Aio` driver in multiple threads and loops environment.

While removing this freedom to our users right now can be seen as none nocive, we could end up in a situation where users have a lot of friction for not having this solved.

Question: What will happen in test scenarios where tests are executed in a new thread? The following issue might appear: Tests running in a different thread could raise a `RuntimeError` exception if the gRPC Asyncio loop was stick to the parent thread.

Reply:
> But at the moment that we are saving at the very beginning which Asyncio loop will be used during the whole application, we explicitly forbidding the pattern for having **many threads and many loops**, , where theoretically the gRPC library might operate - saying theoretically because has been never tested.

I agree. I think supporting "many-thread and many-loop" environment is manageable with the https://github.com/grpc/grpc/pull/22258. However, with custom IO manager, we will need more engineering effort.

To make this library less error-prone, I think you are right that we should imply less restrictions. I created a TODO issue for the support of new environment https://github.com/grpc/grpc/issues/22272. Can we push this PR forward, and solve the many-thread many-loop problem in a new one (it would need test cases and won't be trivial)?

> What will happen in test scenarios where tests are executed in a new thread?

This PR let gRPC pick a thread while initializing async mode. So, if the unit tests are initialized in a new thread, then they should be fine to run. If each test case is executed in a different thread, the exception will be thrown.

Reply:
Thanks!